crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,429 @@
+        return 1;
+    BIO_free(bio);
+#endif
+    CMPerr(0, CMP_R_NO_STDIO);
+    return 0;
+}
+
+void OSSL_CMP_log_close(void) /* is designed to be idempotent */
+{
+    (void)OSSL_trace_set_channel(OSSL_TRACE_CATEGORY_CMP, NULL);
+}
+
+static OSSL_CMP_severity parse_level(const char *level)
+{
+    const char *end_level = strchr(level, ':');
+    int len;
+
+    if (end_level == NULL)
+        return -1;
+
+    if (strncmp(level, OSSL_CMP_LOG_PREFIX,
+                strlen(OSSL_CMP_LOG_PREFIX)) == 0)
+        level += strlen(OSSL_CMP_LOG_PREFIX);
+    len = end_level - level;
+    return
+        strncmp(level, "EMERG", len) == 0 ? OSSL_CMP_LOG_EMERG :
+        strncmp(level, "ALERT", len) == 0 ? OSSL_CMP_LOG_ALERT :
+        strncmp(level, "CRIT", len) == 0 ? OSSL_CMP_LOG_CRIT :
+        strncmp(level, "ERROR", len) == 0 ? OSSL_CMP_LOG_ERR :
+        strncmp(level, "WARN", len) == 0 ? OSSL_CMP_LOG_WARNING :
+        strncmp(level, "NOTE", len) == 0 ? OSSL_CMP_LOG_NOTICE :
+        strncmp(level, "INFO", len) == 0 ? OSSL_CMP_LOG_INFO :
+        strncmp(level, "DEBUG", len) == 0 ? OSSL_CMP_LOG_DEBUG :
+        -1;
+}
+
+const char *ossl_cmp_log_parse_metadata(const char *buf,
+                 OSSL_CMP_severity *level, char **func, char **file, int *line)
+{
+    const char *p_func = buf;
+    const char *p_file = buf == NULL ? NULL : strchr(buf, ':');
+    const char *p_level = buf;
+    const char *prefix = buf;
+
+    *level = -1;
+    *func = NULL;
+    *file = NULL;
+    *line = 0;
+
+    if (p_file != NULL) {
+        const char *p_line = strchr(++p_file, ':');
+
+        /* check if buf contains at least "CMP "followed by logging level */
+        if ((*level = parse_level(buf)) < 0 && p_line++ != NULL) {
+            /* else check if buf contains location info and logging level */
+            char *p_level_tmp = (char *)p_level;
+            const long line_number = strtol(p_line, &p_level_tmp, 10);
+
+            p_level = p_level_tmp;
+            if (p_level > p_line && *(p_level++) == ':') {
+                if ((*level = parse_level(p_level)) >= 0) {
+                    *func = OPENSSL_strndup(p_func, p_file - 1 - p_func);
+                    *file = OPENSSL_strndup(p_file, p_line - 1 - p_file);
+                    *line = (int)line_number;
+                    prefix = p_level;
+                }
+            }
+        }
+    }
+    return prefix;
+}
+
+
+/*
+ * auxiliary function for incrementally reporting texts via the error queue
+ */
+static void put_error(int lib, const char *func, int reason,
+                      const char *file, int line)
+{
+    ERR_new();
+    ERR_set_debug(file, line, func);
+    ERR_set_error(lib, reason, NULL);
+}
+
+#define ERR_print_errors_cb_LIMIT 4096 /* size of char buf2[] variable there */
+#define TYPICAL_MAX_OUTPUT_BEFORE_DATA 100
+#define MAX_DATA_LEN (ERR_print_errors_cb_LIMIT-TYPICAL_MAX_OUTPUT_BEFORE_DATA)
+void OSSL_CMP_add_error_txt(const char *separator, const char *txt)
```

Comment:
Is this still needed...I had a feeling @richsalz did something along these lines already

Reply:
There had been some improvements, but unfortunately the 4096 char buffer limit is still present.
This ought to be either fixed (which shouldn't be too hard) or clearly documented.
I could do this but I don't wanna get this PR blocked even longer.